### PR TITLE
add --expect-error-containing flag

### LINKED ISSUE #1980: Add a plugin flag to signal expected failures
Tests of verification failures currently require to parse the output of compilation failures to check that the errors are the expected ones. If the plugin offered a flag to specify which failures are expected, then the output wouldn't have to be parsed. This would be another step to simplify the testsuite. 

I'm proposing that we add flags
* --expectErrorIncluding "some string" 
* --expectAnyError

The following has verification fail only if no verification failures are found.
```
{-@ LIQUID "--expectAnyError" @-}
```

The following has verification fail only if an error message does not include either letters `a` or `b`.
```
{-@ LIQUID "--expectErrorIncluding \"a\"" @-}
{-@ LIQUID "--expectErrorIncluding \"b\"" @-}
```

In principle, all expected errors are suppressed in the output. This will help the testsuite printing in the output only the errors that aren't expected. We could add later on a flag `--noSuppressExpectedErrors` to have them printed to stderr if needed.

### PR DESCRIPTION
**Pull request by Fizzixnerd**
Closes #1980. 

Still a draft as I finish up.  See above issue for discussion of progress so far.

### CODE REVIEW

File: liquid-base/liquid-base.cabal [hunk 1]
```diff
@@ -257,4 +257,4 @@ library
                       NoImplicitPrelude
   if impl(ghc >= 8.10)
     ghc-options: -fplugin=LiquidHaskell -fplugin-opt=LiquidHaskell:--no-positivity-check
-  ghc-options:        -j
```
Comment (facundominguez) [inline liquid-base/liquid-base.cabal:260]: Is this an unrelated change?

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 2]
```diff
@@ -231,71 +231,74 @@ typecheckHook _ (unoptimise -> modSummary) tcGblEnv = do
...
-
-  pure $ tcGblEnv { tcg_anns = serialisedSpec : tcg_anns tcGblEnv })
-    `gcatch` (\(e :: UserError) -> LH.reportErrors Full [e] >> failM)
-    `gcatch` (\(e :: Error) -> LH.reportErrors Full [e] >> failM)
-    `gcatch` (\(es :: [Error]) -> LH.reportErrors Full es >> failM)
+  let filters = StringFilter <$> expectErrorContaining cfg
+      continue = pure tcGblEnv
+      reportErrs :: (Show e, F.PPrint e) => [TError e] -> TcM TcGblEnv
+      reportErrs = LH.filterReportErrors continue filters Full
+  (do
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs]: Would it be possible to avoid changing the indentation of the code to help the review?

Perhaps it all would be more manageable if the large do block was placed in an auxiliary function.

Also, if `filters` and `reportErrs` had a `cfg` parameter they could be placed in a `where` clause, or at the top level.

File: liquid-base/liquid-base.cabal (hunk 1 above)
Comment (Fizzixnerd) [reply liquid-base/liquid-base.cabal:260]: No, making changes to the pragma code revealed that such flags will crash liquid even when they didn't before.  So I'm removing them everywhere.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs (hunk 2 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/GHC/Plugin.hs]: Changed.

File: src/Language/Haskell/Liquid/UX/CmdLine.hs [hunk 3]
```diff
@@ -742,11 +742,15 @@ reportResult :: MonadIO m
...
              -> [FilePath]
              -> Output Doc
-             -> m ()
+             -> m Bool
 reportResult logResultFull cfg targets out = do
   annm <- {-# SCC "annotate" #-} liftIO $ annotate cfg targets out
   liftIO $ whenNormal $ donePhase Loud "annotate"
-  if | json cfg  -> liftIO $ reportResultJson annm
+  if | json cfg  -> do
+         -- XXX(matt.walker): Does this even work? It does something completely
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/UX/CmdLine.hs]: I don't know what to do here.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 4]
```diff
@@ -72,6 +72,10 @@ import           Language.Haskell.Liquid.UX.CmdLine
 import           Optics
 
 
+data LiquidCheckException =
+  KillPluginWithSuccess
+  deriving (Eq, Ord, Show)
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:84]: I opted for an Either.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs (hunk 2 above)
Comment (facundominguez) [reply src/Language/Haskell/Liquid/GHC/Plugin.hs]: Thanks Matthew. The indentation changes went into a different commit though, which doesn't help observing the diff in the preexisting commit.

Also, the factoring-out of the code to a new function would deserve a standalone commit.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 5]
```diff
@@ -492,7 +492,6 @@ filterReportErrorsWith FilterReportErrorsArgs {..} errs =
     then continue
     else do
       case reducedErrs of
-        -- XXX(matt.walker) there is a law that can simplify this, probably
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs]: Don't know how this got into this commit, but I decided to remove this comment since it was more of a WIP thing when I made it.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 6]
```diff
@@ -261,68 +283,90 @@ liquidHaskellCheck pipelineData modSummary tcGblEnv = (do
...
+liquidHaskellCheck :: PipelineData -> ModSummary -> TcGblEnv -> TcM (Either LiquidCheckException TcGblEnv)
+liquidHaskellCheck pipelineData modSummary tcGblEnv = do
+  cfg <- liftIO getConfig
+  liquidHaskellCheckWithConfig cfg pipelineData modSummary tcGblEnv
+    `gcatch` (\(e :: UserError) -> reportErrs cfg [e])
+    `gcatch` (\(e :: Error) -> reportErrs cfg [e])
+    `gcatch` (\(es :: [Error]) -> reportErrs cfg es)
+  where
+    filters :: Config -> [Filter]
+    filters cfg = getFilters cfg
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs]: Unneeded.

**Comment by Fizzixnerd** [comment]
I rebased on the wrong develop branch sigh.

**Comment by Fizzixnerd** [comment]
We found failing positivity tests that GHC can't compile on cabal_810 even without the plugin: `tests/errors/Positivity1.hs` and `tests/errors/Positivity2.hs`

**Comment by Fizzixnerd** [comment]
> We found failing positivity tests that GHC can't compile on cabal_810 even without the plugin: `tests/errors/Positivity1.hs` and `tests/errors/Positivity2.hs`

Fixed, thanks to @facundominguez

File: tests/errors/ParseBind.hs [hunk 7]
```diff
@@ -1,3 +1,4 @@
+-- Fails because we cannot catch parse errors with --expect-error-containing
 {-@ LIQUID "--expect-error-containing=Cannot parse specification" @-}
 -- We should reject the below to disallow uppercase binders
```
Comment (Fizzixnerd) [inline tests/errors/ParseBind.hs]: Should I fix this?

Comment (Fizzixnerd) [reply tests/errors/ParseBind.hs]: (There is another one in the same commit) 5cd566e

Comment (facundominguez) [reply tests/errors/ParseBind.hs]: Why can't we catch parse errors?

Comment (Fizzixnerd) [reply tests/errors/ParseBind.hs]: A parse error is thrown while parsing the `{-@ @-}` specifications.  So, the `--expect-error-containing` flag is not available when the error is thrown.

Comment (facundominguez) [reply tests/errors/ParseBind.hs]: We could separate and digest the `LIQUID` comments `{-@ LIQUID ... @-}` ahead of parsing the rest of the comments. Could be something for another PR.

Comment (Fizzixnerd) [reply tests/errors/ParseBind.hs]: [7de09d](https://github.com/ucsd-progsys/liquidhaskell/pull/1984/commits/47de09d7b80b66d8ab8006842a546414bccce564) has some similar work that I had to do to catch a bunch of other errors, though what you're suggesting would be slightly more involved, but possible.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 8]
```diff
@@ -477,24 +474,22 @@ data ProcessModuleResult = ProcessModuleResult {
...
-  filters <- getFilters <$> liftIO getConfig
   let commSpecE :: Either [Error] (ModName, Spec LocBareType LocSymbol)
       commSpecE = hsSpecificationP (moduleName thisModule) (coerce specComments) specQuotes
   case commSpecE of
-    Left errors -> do
-      LH.filterReportErrors thisFile GHC.failM continue filters Full errors
+    Left errors ->
+      -- The global config has no filters, so we don't even pretend to check for
+      -- them
+      LH.filterReportErrors thisFile GHC.failM GHC.failM [] Full errors
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs]: This is incredibly wrong.  They can be provided on the command line.

**Comment by Fizzixnerd** [comment]
@facundominguez This PR is ready to start reviewing.  I have some final things to fix up (remove `modify-errors`, make stack work, fix any CI issues, remove newly dead code in `test-driver`), but the core functionality should be there.

File: tests/errors/ParseBind.hs (hunk 7 above)
Comment (Fizzixnerd) [reply tests/errors/ParseBind.hs]: An alternative might be moving those tests to a `parsing-errors` test group, and providing a uniform `--fplugin-opt=LiquidHaskell:--expect-error-containing="Parse"` flag in the cabal file, or similar.  This would provide it without needing to parse the spec.

File: cabal.project [hunk 9]
```diff
@@ -11,6 +11,7 @@ packages: .
           ./tests
           ./tests/benchmarks/popl18/lib
           ./benchmark-timings
+          ./modify-tests
```
Comment (Fizzixnerd) [inline cabal.project]: Remove.

File: modify-tests/CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Revision history for modify-tests
```
Comment (Fizzixnerd) [inline modify-tests/CHANGELOG.md]: Remove.

Comment (Fizzixnerd) [inline modify-tests/CHANGELOG.md]: Remove.

**Comment by Fizzixnerd** [comment]
> @facundominguez This PR is ready to start reviewing. I have some final things to fix up (remove `modify-errors`, make stack work, fix any CI issues, remove newly dead code in `test-driver`), but the core functionality should be there.

I have now finished all of these tasks, except for making cabal 900 work.

File: tests/tests.cabal [hunk 11]
```diff
@@ -17,7 +17,7 @@ flag stack
                " buildable: False " by default
 
 flag measure-timings
-  default:     True
+  default:     False
```
Comment (Fizzixnerd) [inline tests/tests.cabal:20]: Revert.

File: src/Language/Haskell/Liquid/Types/Errors.hs [hunk 12]
```diff
@@ -990,6 +990,10 @@ ppError' _ dCtx (ErrSaved _ name s)
         $+$ dCtx
         $+$ {- nest 4 -} s
 
+-- XXX(matt.walker): If this was because of the limitations taken care of with
+-- the addition of `processPragmas`, then we should remove this. It could be a
+-- "it's too late after the plugin is running to change these values" thing
+-- though, I'm not sure.
```
Comment (Fizzixnerd) [inline src/Language/Haskell/Liquid/Types/Errors.hs:996]: RFC

Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/Errors.hs:996]: cc @nikivazou

File: tests/tests.cabal (hunk 11 above)
Comment (Fizzixnerd) [reply tests/tests.cabal:20]: Actually, this makes sense to keep, no?  It makes the CI run faster it seems because it doesn't have to deal with ten billion lines of output.  Alternatively, I could filter it, but we don't _use_ the timings in the CI currently, so I'm tempted to just disable this flag unless you explicitly ask for it.

**Comment by Fizzixnerd** [comment]
https://github.com/ucsd-progsys/liquidhaskell/pull/1984/commits/45eb0d03ea38a0699c0f53f39b0261ada31bc633 reinstates the infamous liquid header!

**Comment by Fizzixnerd** [comment]
Pass! :tada:

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 13]
```diff
@@ -206,7 +215,7 @@ instance Unoptimise (DynFlags, HscEnv) where
 --    grab from parsing (again) the module by using the GHC API, so we are really
 --    independent from the \"normal\" compilation pipeline.
 --
-typecheckHook :: [CommandLineOption] -> ModSummary -> TcGblEnv -> TcM (Either LiquidCheckException TcGblEnv)
+typecheckHook :: [CommandLineOption] -> ModSummary -> TcGblEnv -> TcM (Either LiquidCheckException (TcGblEnv, [Filter]))
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:222]: Mention in the documentation of the function what the list of filters is.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 14]
```diff
@@ -334,29 +343,31 @@ liquidHaskellCheck pipelineData modSummary tcGblEnv = do
...
+      let bareSpec = lhInputSpec lhContext
+          file = LH.modSummaryHsFile $ lhModuleSummary lhContext
+
+      withPragmas (lhGlobalCfg lhContext) file (Ms.pragmas $ review bareSpecIso bareSpec) $ \moduleCfg ->  do
+        let filters = getFilters moduleCfg
+        -- Report the outcome of the checking
+        wasUnsafeButContinued <- LH.reportResult (errorLogger file filters) moduleCfg [giTarget (giSrc pmrTargetInfo)] out
+        pure $ if wasUnsafeButContinued
+          then Left KillPluginWithSuccess
+          else Right (pmrClientLib, filters)
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:374]: Perhaps the unmatched filters should be preserved when killing the plugin with success.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 15]
```diff
@@ -285,12 +294,12 @@ processInputSpec cfg pipelineData modSummary tcGblEnv inputSpec = do
...
   -- liftIO $ putStrLn ("liquidHaskellCheck 6: " ++ show isIg)
-  liquidLib' :: Either LiquidCheckException LiquidLib <-
+  liquidLib' :: Either LiquidCheckException (LiquidLib, [Filter]) <-
     if isIgnore inputSpec
-    then pure $ Right emptyLiquidLib
+    then pure $ Left KillPluginWithSuccess
     else checkLiquidHaskellContext lhContext
 
-  traverse (serialiseSpec thisModule tcGblEnv) liquidLib'
+  traverse (bitraverse (serialiseSpec thisModule tcGblEnv) pure) liquidLib'
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:306]: Apparently this is evaluated, even if the earlier call to checkLiquidHaskellContext returns KillPluginWithSuccess.

File: tests/errors/Positivity1.hs [hunk 16]
```diff
@@ -4,7 +4,10 @@ module Positivity1 where
 newtype Rec a = In { out :: Rec a -> a }
 
 y :: (a -> a) -> a
-y = \f -> (\x -> f (out x x)) (In (\x -> f (out x x))) 
+y f = g (In g)
+  where
+    {-# NOINLINE g #-}
```
Comment (facundominguez) [inline tests/errors/Positivity1.hs:9]: This is worth a comment.

File: tests/tests.cabal [hunk 17]
```diff
@@ -673,7 +673,11 @@ executable errors
                     , UnboundVarInReflect
                     , UnboundVarInSpec
                     
-    ghc-options:      -fplugin=LiquidHaskell -ddump-timings -ddump-to-file -fkeep-going -O0
+    ghc-options:      -fplugin=LiquidHaskell -fkeep-going -O0
+
+    if flag(measure-timings)
+      ghc-options:    -fforce-recomp -ddump-timings -ddump-to-file
```
Comment (facundominguez) [inline tests/tests.cabal:712]: We should use `measure-timings` in all components.

File: src/Language/Haskell/Liquid/Types/Errors.hs (hunk 12 above)
Comment (facundominguez) [reply src/Language/Haskell/Liquid/Types/Errors.hs:996]: If you are referring to the fix in https://github.com/ucsd-progsys/liquidhaskell/pull/1984/commits/ecbc2fda203b9b034cb72fd6e60ec1509917e453, it looks to me like this should be unrelated. I would remove this comment.

File: tests/datacon/neg/Data00.hs [hunk 18]
```diff
@@ -1,4 +1,3 @@
-{-@ LIQUID "--expect-any-error" @-}
```
Comment (facundominguez) [inline tests/datacon/neg/Data00.hs]: I don't understand these removals. This is a neg test but it produces no expected errors?

File: tests/tests.cabal (hunk 11 above)
Comment (facundominguez) [reply tests/tests.cabal:20]: It makes sense to me to disable it for now. But then you need to explain in the readme how to enable it to collect timings.

File: tests/tests.cabal [hunk 19]
```diff
@@ -583,9 +613,6 @@ executable errors
                     , BadDataConType
                     , BadDataDeclTyVars
                     , BadGADT
-                    -- , BadPragma0
-                    -- , BadPragma1
-                    -- , BadPragma2
```
Comment (facundominguez) [inline tests/tests.cabal:586]: There are various disabled tests that are removed in this file. Are they beyond any hope of enabling them again?

**Comment by facundominguez** [comment]
Thanks @Fizzixnerd! It is looking good. Please, when disabling tests, create an issue to investigate them and enable them back.

**Comment by facundominguez** [comment]
Another thing to do before merging is checking that the instructions in the README stay up-to-date for running tests.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 20]
```diff
@@ -127,9 +134,18 @@ plugin = GHC.defaultPlugin {
...
           liftIO $ printWarning dynFlags warning
           pure gblEnv
-        else typecheckHook opts summary gblEnv 
-
-
+        else do
+          newGblEnv <- typecheckHook opts summary gblEnv
+          case newGblEnv of
+            Left KillPluginWithSuccess -> pure gblEnv
+            Right (newGblEnv', errorFilters) ->
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:145]: The convention in Haskell is to use `Left` to report errors. 
Maybe `LiquidCheckException` could have another case to report filters that haven't been matched.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 21]
```diff
@@ -70,7 +73,11 @@ import           Language.Haskell.Liquid.Bare
 import           Language.Haskell.Liquid.UX.CmdLine
 
 import           Optics
+import           Data.Bitraversable                       (bitraverse)
 
+data LiquidCheckException =
+  KillPluginWithSuccess
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs:83]: This is a fantastic place to explain the meaning of `LiquidCheckException`, i.e. how it differs from other errors, and explain when is `KillPluginWithSuccess` delivered.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs (hunk 15 above)
Comment (facundominguez) [reply src/Language/Haskell/Liquid/GHC/Plugin.hs:306]: We agreed during the review that this code is correct, but non-obvious, and needs to be changed to make the control flow explicit.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 22]
```diff
@@ -62,6 +72,10 @@ import           Language.Haskell.Liquid.Types.Types
 import           Prelude                          hiding (error)
 import           Text.PrettyPrint.HughesPJ        hiding ((<>))
 
+data Filter = StringFilter String
+            | AnyFilter
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs:80]: This is a good place to explain what a Filter is, where they come from, and how they are used.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 23]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+data FilterReportErrorsArgs m filter msg e a =
+  FilterReportErrorsArgs
+  { msgReporter :: [msg] -> m ()
+  , filterReporter :: [filter] -> m ()
+  , failure :: m a
+  , continue :: m a
+  , pprinter :: e -> m msg
+  , filterMapper :: e -> Either [e] [filter]
+  , filters :: [filter]
+  }
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs:524]: Please, document the meaning of the fields.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 24]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+
+-- | Abstract filtered error reporter
+filterReportErrorsWith :: (Monad m, Ord filter) => FilterReportErrorsArgs m filter msg e a -> [e] -> m a
+filterReportErrorsWith FilterReportErrorsArgs {..} errs =
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs:528]: Please, explain what the function is supposed to do. The description is insufficient to understand the properties that are being tested in the testsuite.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 25]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+-- | Retrieve the `Filter`s from the Config.
+getFilters :: Config -> [Filter]
+getFilters cfg = anyFilter <> stringFilters
+  where
+    anyFilter = [AnyFilter | expectAnyError cfg]
+    stringFilters = StringFilter <$> expectErrorContaining cfg
+
+-- | Return either the error as a singleton or the list of @filters@ that
+-- matched the @err@, given a @renderer@ for the @err@ and some @filters@
+reduceFilters :: forall e. (e -> String) -> [Filter] -> e -> Either [e] [Filter]
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs]: This type looks more complicated than necessary. How is it better than returning `[Filter]` and allowing the result to be empty?

It looks to me like this is transferring concerns from the call site to the function, which obscures its meaning. I think that would deserve to be reverted, even if it means making the call site a bit more complicated.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 26]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+
+-- | Retrieve the `Filter`s from the Config.
+getFilters :: Config -> [Filter]
+getFilters cfg = anyFilter <> stringFilters
+  where
+    anyFilter = [AnyFilter | expectAnyError cfg]
+    stringFilters = StringFilter <$> expectErrorContaining cfg
+
+-- | Return either the error as a singleton or the list of @filters@ that
+-- matched the @err@, given a @renderer@ for the @err@ and some @filters@
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs]: Crucially, the description of the function omits saying when Right or Left are returned. Please, could we make this explicit?

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 27]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+          msgs <- traverse pprinter es
+          void $ msgReporter msgs
+        Right fs -> do
+          let missedFilters = Set.toList $ Set.fromList filters `Set.difference` Set.fromList fs
+          filterReporter missedFilters
+      failure
+
+-- | Report errors via GHC's API stating the given `Filter`s did not get
+-- matched.
+defaultFilterReporter :: FilePath -> [Filter] -> Ghc.TcRn ()
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs:550]: Either do nothing on an empty list, or ask the list to be non-empty and have the code throw an error if violated.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs [hunk 28]
```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
...
+    wasExactlyFiltered =
+      case reducedErrs of
+        Left es -> null es
+        Right fs -> Set.fromList filters == Set.fromList fs
+  in
+    if wasExactlyFiltered
+    then continue
+    else do
+      case reducedErrs of
+        Left es -> do
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/PrettyPrint.hs:540]: This is going to report only one unmatched error (in a singleton list), when ideally all unmatched errors should be reported. The traverse call used above does not make this particularly obvious either.

**Comment by facundominguez** [comment]
One other thing to do is either suppressing warnings or fixing them in the tests, in order to get a clear output when there are errors.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs (hunk 27 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/PrettyPrint.hs:550]: Changed.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs (hunk 26 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/PrettyPrint.hs]: Changed.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs (hunk 25 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/PrettyPrint.hs]: Changed.

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs (hunk 23 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/PrettyPrint.hs:524]: Changed, but with a TODO

File: src/Language/Haskell/Liquid/Types/PrettyPrint.hs (hunk 22 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/Types/PrettyPrint.hs:80]: Changed.

File: src/Language/Haskell/Liquid/GHC/Plugin.hs (hunk 21 above)
Comment (Fizzixnerd) [reply src/Language/Haskell/Liquid/GHC/Plugin.hs:83]: Changed.

**Comment by facundominguez** [comment]
I'm closing this to reopen it on a branch that I can edit.
